File: pnv_person.rb
class Pnv_person

    @@categories = [
	"prefix",
	"givenName",
	"patronym",
	"givenNameSuffix",
	"infixTitle",
	"surnamePrefix",
	"baseSurname",
	"trailingPatronym",
	"honorificSuffix",
	"disambiguatingDescription"
    ]

    @@combinations = [
	"surname",
	"literalName",
	"firstName",
	"infix",
	"suffix"
    ]

    @naam = Hash.new

    def initialize
    	@naam = Hash.new
    end
    
    def add categorie,value
    	if @naam[categorie].nil?
    	    @naam[categorie] = value
    	else
    	    @naam[categorie] += " #{value}"
    	    STDERR.puts "added #{categorie} more than once (#{@naam[categorie]})"
    	end
    	STDERR.puts "#{categorie} is unknown as a name part" if !@@categories.include? categorie
    end

    def get categorie
    	return combine categorie if @@combinations.include? categorie
    	return @naam[categorie] unless @@categories.include? categorie
    	return "unknown name part: #{categorie}"
    end

    def combine categorie
    	if categorie.eql?("literalName")
    	    result = ""
    	    @@categories.each do |cat|
        		result += " #{@naam[cat]}"
	        end
	        return clean result
	    elsif categorie.eql?("surname")
	        return clean "#{@naam['surnamePrefix']} #{@naam['baseSurname']}"
	    elsif categorie.eql?("first")
	        return clean "#{@naam['givenName']} #{@naam['patronym']} #{@naam['givenNameSuffix']}"
	    elsif categorie.eql?("infix")
	        return clean "#{@naam['InfixTitle']} #{@naam['surnamePrefix']}"
	    elsif categorie.eql?("suffix")
	        return clean "#{@naam['tralingPatronym']} #{@naam['honorificSuffix']} #{@naam['disambiguatingDescription']}"
	    end
    end

    def clean text
	    text.gsub(/  +/," ").gsub(/' /,"'").strip
    end
    
    def to_h
	    return @naam
    end

    def to_s
	    res = "{\n"
	    @naam.each do |k,v|
	        res += "  #{k} => #{v},\n"
    	end
    	res.strip!
    	res = res[0..-2] if res[-1..-1].eql?(',')
    	res += "}"
    	return res
    end

end

File: pnv_person.py
# -*- coding: utf-8 -*-
import re
import sys

class PnvPerson(dict):
    def __missing__(self, key):
        return ""

    categories = [
        "prefix",
        "givenName",
        "patronym",
        "givenNameSuffix",
        "infixTitle",
        "surnamePrefix",
        "baseSurname",
        "trailingPatronym",
        "honorificSuffix",
        "disambiguatingDescription"
    ]

    combinations = [
        "surname",
        "literalName",
        "firstName",
        "infix",
        "suffix"
    ]

    def __init__(self):
        self.naam = {}

    def add(self,categorie,value):
        if categorie in self.naam:
            self.naam[categorie] += " {}".format(value)
        else:
            self.naam[categorie] = value

    def get(self,categorie):
        if categorie in self.combinations:
            return self.combine(categorie) 
        if categorie not in self.categories:
            try:
                return self.naam[categorie]
            except KeyError:
                pass
        return "unknown name part: {}".format(categorie)

    def __get(self, categorie):
        try:
            return self.naam[categorie]
        except KeyError:
            return ""

    def combine(self,categorie):
        if(categorie == "literalName"):
            result = ""
            for cat in self.categories:
                if cat in self.naam:
                    result = result + " {}".format(self.naam[cat])
            return self.clean(result)
        elif(categorie == "surname"):
            return self.clean(f"{self.naam['surnamePrefix']} {self.naam['baseSurname']}")
        elif(categorie == "firstName"):
            return self.clean(f"{self.__get('givenName')} {self.__get('patronym')} {self.__get('givenNameSuffix')}")
        elif(categorie == "infix"):
            return self.clean(f"{self.__get('InfixTitle')} {self.__get('surnamePrefix')}")
        elif(categorie == "suffix"):
            return self.clean(f"{self.__get('trailingPatronym')} {self.__get('honorificSuffix')} {self.__get('disambiguatingDescription')}")

    def clean(self,text):
        pattern = re.compile("  +")
        pattern_2 = re.compile("' ")
        result = re.sub(pattern," ",text)
        result = re.sub(pattern_2,"'",result)
        return result.strip()

    def to_h(self):
        return self.naam

    def to_s(self):
        return self.get('literalName')

